### PR TITLE
Added positions to all castle blocks

### LINKED ISSUE #2: No Position on Building Blocks
Would be nice if the blocks that create castle components (like build tower) also took in a location. Could hide behind a + if needed.

### PR DESCRIPTION
**Pull request by srietkerk**
image

Fixes https://github.com/microsoft/makecode-minecraft-castle-builder/issues/2

### CODE REVIEW

**Review by abchatra — COMMENTED** [review]
The relative position to camera may be easier for students than the relative position to player.
image

File: castles.ts [hunk 1]
```diff
@@ -17,10 +17,21 @@ namespace castles {
...
-        builder.shift(-Math.floor(width / 2), 0, -Math.floor(width / 2))
+    export function buildCastleWall(
+        width: number,
+        length: number,
+        height: number,
+        blockType: number,
+        position?: Position
+        ) {
+        if (position) {
+            builder.shift(position.getValue(Axis.X) - Math.floor(width / 2), 0, position.getValue(Axis.Z) - Math.floor(width / 2))
```
Comment (kimprice) [inline castles.ts]: Are you missing the Y value?

File: castles.ts [hunk 2]
```diff
@@ -37,18 +48,23 @@ namespace castles {
...
     //% blockType.defl=Block.Cobblestone
     //% inlineInputMode="external"
+    //% position.shadow=minecraftCreatePosition
     //% weight=350
     //% help=github:makecode-minecraft-castle-builder/docs/castle-tower
-    export function buildCastleTower(width: number, height: number, blockType: number) {
-        builder.shift(- Math.floor(width / 2), 0, - Math.floor(width / 2))
+    export function buildCastleTower(width: number, height: number, blockType: number, position?: Position) {
+        if (position) {
+            builder.shift(position.getValue(Axis.X) - Math.floor(width / 2), 0, position.getValue(Axis.Z) - Math.floor(width / 2))
```
Comment (kimprice) [inline castles.ts]: Missing the Y value here too?

File: castles.ts (hunk 1 above)
Comment (srietkerk) [reply castles.ts]: Good catch, thank you!!!

**Comment by srietkerk** [comment]
> The relative position to camera may be easier for students than the relative position to player. image

@abchatra I'll do this because I agree, it's a more readable position type, but I just wanted to note that the field editor for this type of position maxes the sliders out at -8 and 8. That said, a user can just _type_ something into the number fields that is greater than 8 and it works fine, but I just wanted to point this out as a possible pain point. I chatted with Joey, and he showed me how I can make these values larger with a new function that defines minimums and maximums for the field editor, but this has to be done in pxt-minecraft first. I wanted to bring this up because even though I don't think making this new block will be much work at all, I'm still wondering if it is something I should work on or if I should leave the field editor with the max values of 8 and see if there is any feedback from users.
image

**Comment by jwunderl** [comment]
> I chatted with Joey, and he showed me how I can make these values larger with a new function that defines minimums and maximums for the field editor, but this has to be done in pxt-minecraft first. 

This should be fine to do in the extension itself actually, it's in the 'user level' code so same level of authoring as extension (that is, it's in `/libs` not `/sim` https://github.com/microsoft/pxt-minecraft/blob/master/libs/core/helpers.ts#L93, assuming I didn't have a typo in the field editor itself or something)

**Comment by srietkerk** [comment]
> This should be fine to do in the extension itself actually, it's in the 'user level' code so same level of authoring as extension (that is, it's in `/libs` not `/sim` https://github.com/microsoft/pxt-minecraft/blob/master/libs/core/helpers.ts#L93, assuming I didn't have a typo in the field editor itself or something)

Is this preferred to putting it in pxt-minecraft? Putting the custom function here will certainly make my life simpler, but if we want to use this for other extensions or projects, it might be nice to have it in a common place.

**Review by kimprice — APPROVED** [review]
LGTM! I know you're still waiting on changing the slider range, though.

**Comment by srietkerk** [comment]
> LGTM! I know you're still waiting on changing the slider range, though.

I actually think I'm going to wait on increasing the slider range. I'll make an issue for it just to keep track of it and point out that it's something to consider, but being that it doesn't sound like users have a problem with it by itself, I might just be trying to make more work than necessary.

**Pull request merged**